### PR TITLE
Enforce SIMD alignment during index creation

### LINKED ISSUE #565: Index creation time can be slow when there's no SIMD alignment
In `LanceDataset.create_index`,  if (num_partitions / num_subvectors) % 8 != 0  then there is no SIMD alignment so index creation speed becomes interminably slow.

The user does not know to do that or have information to debug this when it's the case. We should:

1. If the user does not specify `num_sub_vectors`, automatically calculate an optimal one that guarantees SIMD alignment. If the number of dimensions is not a multiple of 8, then print a warning message.
2. If the user does specify `num_sub_vectors`, check if the quotient above is a multiple of 8 and print warning message as necessary.

### PR DESCRIPTION
**Pull request by changhiskhan**
Closes #565 

Assert that ndim / num_sub_vector is a multiple of the SIMD stride (8 on ubuntu and 4 on arm)

Raise error if the alignment is off

### CODE REVIEW

File: rust/src/dataset.rs [hunk 1]
```diff
@@ -508,6 +529,14 @@ impl Dataset {
     }
 }
 
+#[inline]
+fn simd_stride() -> i32 {
```
Comment (eddyxu) [inline rust/src/dataset.rs:533]: Put into a separate module? and call it `simd_alignment`, which is consistent with the rest of Rust toolchain. https://doc.rust-lang.org/reference/type-layout.html#size-and-alignment

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -251,11 +251,15 @@ def create_index(
             Parameters passed to the index building process.
 
 
-        Accepted keyword parameters:
+        If `index_type` is "IVF_PQ", then the following parameters are required:
 
         - **num_partitions**: the number of partitions of IVF (Inverted File Index).
         - **num_sub_vectors**: the number of sub-vectors used in Product Quantization.
 
+        For SIMD, the vector dimensions / num_sub_vectors must be a multiple of the stride
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/python/lance/dataset.py:259]: added

File: python/src/dataset.rs [hunk 3]
```diff
@@ -204,17 +204,21 @@ impl Dataset {
         let mut params = VectorIndexParams::default();
         if let Some(n) = kwargs.get_item("num_partitions") {
             params.num_partitions = PyAny::downcast::<PyInt>(n)?.extract()?
+        } else {
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/src/dataset.rs]: moved to python

File: rust/src/dataset.rs (hunk 1 above)
Comment (changhiskhan) [reply rust/src/dataset.rs]: put it in distance for now since that's what needs the check

**Review by eddyxu — APPROVED** [review]

File: rust/src/utils/distance.rs [hunk 4]
```diff
@@ -26,6 +26,18 @@ use arrow_schema::DataType;
 
 use crate::Result;
 
+#[inline]
+pub(crate) fn simd_alignment() -> i32 {
+    #[cfg(any(target_arch = "x86_64"))]
+    {
```
Comment (eddyxu) [inline rust/src/utils/distance.rs:32]: this should be `AVX256` -> 8, `AVX512` -> 16, tho, and `neon` -> 4. 

We can change it later if we decide to add AVX512 support

**Pull request merged**